Data prep purchase orders


*** units: NaN, PUND and KG -> transform all to KG

*** what to do with negative values

*** date_time columns have various time zones (+00:00, +01:00, +02:00)

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [3]:
purchase = pd.read_csv("/home/celinaha/ml_group21/data/kernel/purchase_orders.csv").copy()

In [6]:
purchase.head()

purchase_order_id  purchase_order_item_no  quantity  \
0                  1                       1     -14.0   
1                 22                       1   23880.0   
2                 41                       1       0.0   
3                 61                       1       0.0   
4                141                      10   25000.0   

                        delivery_date  product_id  product_version  \
0  2003-05-12 00:00:00.0000000 +02:00    91900143                1   
1  2003-05-27 00:00:00.0000000 +02:00    91900160                1   
2  2004-03-08 00:00:00.0000000 +01:00    91900143                1   
3  2004-03-10 00:00:00.0000000 +01:00    91900143                1   
4  2004-10-28 00:00:00.0000000 +02:00    91900143                1   

                    created_date_time                  modified_date_time  \
0  2003-05-12 10:00:48.0000000 +00:00  2004-06-15 06:16:18.0000000 +00:00   
1  2003-05-27 12:42:07.0000000 +00:00  2012-06-29 09:41:13.0000000 +00:00   
2  2004-03-08 13:44:31.0000000 +00:00  2012-07-04 13:51:02.0000000 +00:00   
3  2004-03-10 11:39:06.0000000 +00:00  2012-07-04 13:50:59.0000000 +00:00   
4  2004-10-22 12:21:54.0000000 +00:00  2012-07-04 13:50:55.0000000 +00:00   

   unit_id unit  status_id  status  
0      NaN  NaN          2  Closed  
1      NaN  NaN          2  Closed  
2      NaN  NaN          2  Closed  
3      NaN  NaN          2  Closed  
4      NaN  NaN          2  Closed

In [4]:
purchase.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 33171 entries, 0 to 33170
Data columns (total 12 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   purchase_order_id       33171 non-null  int64  
 1   purchase_order_item_no  33171 non-null  int64  
 2   quantity                33171 non-null  float64
 3   delivery_date           33171 non-null  object 
 4   product_id              33171 non-null  int64  
 5   product_version         33171 non-null  int64  
 6   created_date_time       33171 non-null  object 
 7   modified_date_time      32679 non-null  object 
 8   unit_id                 33127 non-null  float64
 9   unit                    33127 non-null  object 
 10  status_id               33171 non-null  int64  
 11  status                  33171 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 3.0+ MB


In [5]:
purchase.nunique()

purchase_order_id          8135
purchase_order_item_no       94
quantity                   4997
delivery_date              2332
product_id                   77
product_version               9
created_date_time         14657
modified_date_time        22845
unit_id                       2
unit                          2
status_id                     3
status                        3
dtype: int64

In [15]:
purchase['unit'].unique()

array([nan, 'KG', 'PUND'], dtype=object)

Removing duplicates:

In [21]:
print(purchase.duplicated().any())

False


Fixing DateTime, and Date:

In [30]:
purchase['created_date_time'] = pd.to_datetime(purchase['created_date_time'], errors='coerce')
purchase['modified_date_time'] = pd.to_datetime(purchase['modified_date_time'], errors='coerce')

In [31]:
purchase['delivery_date'] = pd.to_datetime(purchase['delivery_date'], errors='coerce')


In [ ]:
purchase.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 33171 entries, 0 to 33170
Data columns (total 12 columns):
 #   Column                  Non-Null Count  Dtype                    
---  ------                  --------------  -----                    
 0   purchase_order_id       33171 non-null  int64                    
 1   purchase_order_item_no  33171 non-null  int64                    
 2   quantity                33171 non-null  float64                  
 3   delivery_date           19132 non-null  datetime64[ns, UTC+02:00]
 4   product_id              33171 non-null  int64                    
 5   product_version         33171 non-null  int64                    
 6   created_date_time       33171 non-null  datetime64[ns, UTC]      
 7   modified_date_time      32679 non-null  datetime64[ns, UTC]      
 8   unit_id                 33127 non-null  float64                  
 9   unit                    33127 non-null  object                   
 10  status_id               33171 non-

Fixing different timezones:

Removing negative values:

In [9]:
(purchase.select_dtypes(include=['number']) < 0).any()

purchase_order_id         False
purchase_order_item_no    False
quantity                   True
product_id                False
product_version           False
unit_id                   False
status_id                 False
dtype: bool

In [10]:
purchase[purchase['quantity']<0]

purchase_order_id  purchase_order_item_no  quantity  \
0                    1                       1     -14.0   
59              205323                      10    -380.0   
60              205443                      10    -390.0   
61              206151                      10    -260.0   
66              206386                      10    -340.0   
208             209842                      10   -8580.0   

                          delivery_date  product_id  product_version  \
0    2003-05-12 00:00:00.0000000 +02:00    91900143                1   
59   2003-01-31 00:00:00.0000000 +01:00        1024                1   
60   2003-02-10 00:00:00.0000000 +01:00        1024                1   
61   2003-04-21 00:00:00.0000000 +02:00        1024                1   
66   2003-05-14 00:00:00.0000000 +02:00        1024                1   
208  2004-04-30 00:00:00.0000000 +02:00    91900170                1   

                      created_date_time                  modified_date_time  \
0    2003-05-12 10:00:48.0000000 +00:00  2004-06-15 06:16:18.0000000 +00:00   
59   2003-01-28 13:30:23.0000000 +00:00  2004-06-15 06:08:40.0000000 +00:00   
60   2003-02-10 09:31:21.0000000 +00:00  2004-06-15 06:14:46.0000000 +00:00   
61   2003-04-14 09:31:14.0000000 +00:00  2004-06-15 06:14:50.0000000 +00:00   
66   2003-05-09 08:30:53.0000000 +00:00  2004-06-15 06:14:54.0000000 +00:00   
208  2004-04-28 08:32:31.0000000 +00:00  2004-06-15 06:16:36.0000000 +00:00   

     unit_id unit  status_id  status  
0        NaN  NaN          2  Closed  
59      40.0   KG          2  Closed  
60      40.0   KG          2  Closed  
61      40.0   KG          2  Closed  
66      40.0   KG          2  Closed  
208     40.0   KG          2  Closed

Missing values; handling NaN in unit

In [20]:
purchase.isna().sum()

purchase_order_id           0
purchase_order_item_no      0
quantity                    0
delivery_date               0
product_id                  0
product_version             0
created_date_time           0
modified_date_time        492
unit_id                    44
unit                       44
status_id                   0
status                      0
dtype: int64

In [19]:
purchase[purchase['unit']=='PUND']

purchase_order_id  purchase_order_item_no     quantity  \
20597             274174                      10  1079.998444   
21398             277831                      10  1079.998444   
21583             278932                      10  1079.998444   
21906             279515                      10  1079.998444   
22121             280240                      10  2159.996888   
22666             282355                      10  2159.996888   
23817             286407                      10  1079.998444   

                            delivery_date  product_id  product_version  \
20597  2015-06-25 00:00:00.0000000 +02:00        1414                2   
21398  2016-01-19 00:00:00.0000000 +01:00        1414                2   
21583  2016-03-08 00:00:00.0000000 +01:00        1414                2   
21906  2016-04-14 00:00:00.0000000 +02:00        1414                2   
22121  2016-05-17 00:00:00.0000000 +02:00        1414                2   
22666  2016-09-20 00:00:00.0000000 +02:00        1414                2   
23817  2017-04-28 00:00:00.0000000 +02:00        1414                2   

                        created_date_time modified_date_time  unit_id  unit  \
20597  2015-06-16 13:10:15.0000000 +00:00                NaN     43.0  PUND   
21398  2016-01-14 10:51:26.0000000 +00:00                NaN     43.0  PUND   
21583  2016-03-01 15:15:10.0000000 +00:00                NaN     43.0  PUND   
21906  2016-04-07 10:29:09.0000000 +00:00                NaN     43.0  PUND   
22121  2016-05-11 15:27:56.0000000 +00:00                NaN     43.0  PUND   
22666  2016-09-13 15:23:26.0000000 +00:00                NaN     43.0  PUND   
23817  2017-04-25 14:20:17.0000000 +00:00                NaN     43.0  PUND   

       status_id status  
20597          1   Open  
21398          1   Open  
21583          1   Open  
21906          1   Open  
22121          1   Open  
22666          1   Open  
23817          1   Open

*** Product id 1414 has units pund

After fixing datetime:

In [35]:
purchase.isna().sum()

purchase_order_id             0
purchase_order_item_no        0
quantity                      0
delivery_date             14039
product_id                    0
product_version               0
created_date_time             0
modified_date_time          492
unit_id                      44
unit                         44
status_id                     0
status                        0
dtype: int64

In [38]:
purchase[purchase['delivery_date'].isna()]

purchase_order_id  purchase_order_item_no  quantity delivery_date  \
2                     41                       1       0.0           NaT   
3                     61                       1       0.0           NaT   
5                    161                      10    6000.0           NaT   
35                   468                      10    1000.0           NaT   
36                   469                      10    1000.0           NaT   
...                  ...                     ...       ...           ...   
33166             331152                      10  100000.0           NaT   
33167             331202                      10   50000.0           NaT   
33168             331202                      20   50000.0           NaT   
33169             331202                      30   50000.0           NaT   
33170             331202                      40  100000.0           NaT   

       product_id  product_version         created_date_time  \
2        91900143                1 2004-03-08 13:44:31+00:00   
3        91900143                1 2004-03-10 11:39:06+00:00   
5        91900143                1 2005-03-11 13:53:25+00:00   
35       91901460                1 2020-10-01 11:34:47+00:00   
36       91901460                1 2020-10-01 11:38:32+00:00   
...           ...              ...                       ...   
33166    91901460                1 2024-12-20 11:41:29+00:00   
33167    91901460                1 2024-12-27 09:53:18+00:00   
33168    91901460                1 2024-12-27 09:53:17+00:00   
33169    91901460                1 2024-12-27 09:53:18+00:00   
33170    91901460                1 2024-12-27 09:53:18+00:00   

             modified_date_time  unit_id unit  status_id  status  
2     2012-07-04 13:51:02+00:00      NaN  NaN          2  Closed  
3     2012-07-04 13:50:59+00:00      NaN  NaN          2  Closed  
5     2012-07-04 13:50:49+00:00      NaN  NaN          2  Closed  
35    2021-02-26 11:07:07+00:00      NaN  NaN          2  Closed  
36    2021-02-26 10:59:54+00:00      NaN  NaN          2  Closed  
...                         ...      ...  ...        ...     ...  
33166 2025-02-19 12:57:21+00:00     40.0   KG          2  Closed  
33167 2025-02-19 12:57:22+00:00     40.0   KG          2  Closed  
33168 2025-02-19 12:57:22+00:00     40.0   KG          2  Closed  
33169 2025-02-19 12:57:22+00:00     40.0   KG          2  Closed  
33170 2025-02-19 12:57:22+00:00     40.0   KG          2  Closed  

[14039 rows x 12 columns]